src/history/hs.c

```diff
@@ -406,17 +387,65 @@ __hs_insert_record_with_btree(WT_SESSION_IMPL *session, WT_CURSOR *cursor, WT_BT
          */
         WT_TRET(__wt_cursor_key_order_init(cbt));
 #endif
-        WT_TRET(__hs_delete_key(session, cursor, btree->id, key));
+        session_flags = session->flags;
+        F_SET(session, WT_SESSION_IGNORE_HS_TOMBSTONE);
+        /* We're pointing at the newly inserted update. Iterate once more to avoid deleting it. */
+        ret = cursor->next(cursor);
+        if (ret == WT_NOTFOUND)
+            ret = 0;
+        else if (ret == 0)
+            WT_TRET(__hs_delete_key_from_pos(session, cursor, btree->id, key));
+        if (!FLD_ISSET(session_flags, WT_SESSION_IGNORE_HS_TOMBSTONE))
+            F_CLR(session, WT_SESSION_IGNORE_HS_TOMBSTONE);
     }
     /* We did a row search, release the cursor so that the page doesn't continue being held. */
     cursor->reset(cursor);
 
-    /* The tree structure can change while we try to insert the mod list, retry if that happens. */
-    if (ret == WT_RESTART)
-        goto retry;
+    return (ret);
+}
 
-    if (ret != 0)
-        __wt_free_update_list(session, &hs_upd);
+/*
+ * __hs_insert_record_with_btree --
+ *     A helper function to insert the record into the history store including stop time pair.
+ *     Should be called with session's btree switched to the history store.
+ */
+static int
+__hs_insert_record_with_btree(WT_SESSION_IMPL *session, WT_CURSOR *cursor, WT_BTREE *btree,
+  const WT_ITEM *key, const WT_UPDATE *upd, const uint8_t type, const WT_ITEM *hs_value,
+  WT_TIME_PAIR stop_ts_pair)
+{
+    WT_DECL_RET;
+
+    /*
```

Comment:
Add assert to check session's btree is HS.

Reply:
24597ee